paritydb/src/transaction.rs

```
@@ -57,34 +58,65 @@ impl<'a> Operation<'a> {
 
 /// Database operations.
 pub struct Transaction {
+	/// key length, it's used to determine whether an insert
+	/// is valid or not at an early stage, we could probably
+	/// use `Option` or `InternalOption` here, but right now
+	/// we only care about key size, so it's enough info.
+	key_len: usize,
+	first_invalid_key_len: Option<usize>,
 	operations: Vec<u8>,
 }
 
-impl Default for Transaction {
-	fn default() -> Self {
+impl Transaction {
+	/// This should only be called in `Database` and some unit tests.
+	/// Use `db.create_transaction()` in any other cases.
+	pub(crate) fn new(key_len: usize) -> Transaction {
 		Transaction {
+			key_len: key_len,
+			first_invalid_key_len: None,
 			operations: Vec::new(),
 		}
 	}
-}
 
-impl Transaction {
 	/// Append new insert operation to the list of transactions.
 	#[inline]
 	pub fn insert<K: AsRef<[u8]>, V: AsRef<[u8]>>(&mut self, key: K, value: V) {
-		self.push(Operation::Insert(key.as_ref(), value.as_ref()));
+		if self.first_invalid_key_len.is_some() {
+			// there's already something wrong here
+			return;
+		}
+
+		let key = key.as_ref();
+		if key.len() != self.key_len {
+			self.first_invalid_key_len = Some(key.len());
+		} else {
+			self.push(Operation::Insert(key, value.as_ref()));
+		}
 	}
 
 	/// Append new delete operation to the list of transactions.
 	#[inline]
 	pub fn delete<K: AsRef<[u8]>>(&mut self, key: K) {
-		self.push(Operation::Delete(key.as_ref()));
+		if self.first_invalid_key_len.is_some() {
+			// there's already something wrong here
+			return;
+		}
+
+		let key = key.as_ref();
+		if key.len() != self.key_len {
+			self.first_invalid_key_len = Some(key.len());
+		} else {
+			self.push(Operation::Delete(key));
+		}
 	}
 
 	/// Returns double-ended iterator over all operations in a transaction.
-	pub fn operations(&self) -> OperationsIterator {
-		OperationsIterator {
-			data: &self.operations,
+	pub fn operations(&self) -> Result<OperationsIterator> {
```

Why not return `Result` from `insert()` instead? I think I would rather see it failing early (might help in detecting bugs).

---

Sorry for being late, yes, I agree that returning errors as early as possible is nice and I'll change that.